src/Map.ts

```diff
@@ -1190,4 +1208,209 @@ export class Map extends maplibregl.Map {
     super.setTransformRequest(combineTransformRequest(transformRequest));
     return this;
   }
+
+  /**
+   * Add a polyline to the map from various source and with builtin styling.
+   * Compatible sources:
+   * - gpx content as string
+   * - gpx file from URL
+   * - kml content from string
+   * - kml from url
+   * - geojson from url
+   * - geojson content as string
+   * - geojson content as JS object
+   * - uuid of a MapTiler Cloud dataset
+   *
+   * The method also gives the possibility to add an outline layer (if `options.outline` is `true`)
+   * and if so , the returned property `polylineOutlineLayerId` will be a string. As a result, two layers
+   * would be added.
+   *
+   * The default styling creates a line layer of constant width of 3px, the color will be randomly picked
+   * from a curated list of colors and the opacity will be 1.
+   * If the outline is enabled, the outline width is of 1px at all zoom levels, the color is white and
+   * the opacity is 1.
+   *
+   * Those style properties can be changed and ramped according to zoom level using an easier syntax.
+   *
+   * @param options
+   * @param fetchOptions
+   * @returns
+   */
+  async addPolyline(
+    options: PolylineLayerOptions,
+    fetchOptions: RequestInit = {},
+  ): Promise<{
+    polylineLayerId: string;
+    polylineOutlineLayerId: string | null;
+    polylineSourceId: string;
+  }> {
+    // We need to have the sourceId of the sourceData
+    if (!options.sourceId && !options.data) {
+      throw new Error(
+        "Creating a polyline layer requires an existing .sourceId or a valid .data property",
+      );
+    }
+
+    // We are going to evaluate the content of .data, if provided
+    let data = options.data as any;
+    let tmpData = null;
+
+    if (typeof data === "string") {
+      // if options.data exists and is a uuid string, we consider that it points to a MapTiler Dataset
+      if (isUUID(data)) {
+        data = `https://api.maptiler.com/data/${options.data}/features.json?key=${config.apiKey}`;
+      }
+
+      // options.data could be a url to a .gpx file
+      else if (data.split(".").pop()?.toLowerCase().trim() === "gpx") {
+        // fetch the file
+        const res = await fetch(data, fetchOptions);
+        const gpxStr = await res.text();
+        // Convert it to geojson. Will throw is invalid GPX content
+        data = gpx(gpxStr) as FeatureCollection;
+      }
+
+      // options.data could be a url to a .kml file
+      else if (data.split(".").pop()?.toLowerCase().trim() === "kml") {
+        // fetch the file
+        const res = await fetch(data, fetchOptions);
+        const kmlStr = await res.text();
+        // Convert it to geojson. Will throw is invalid GPX content
+        data = kml(kmlStr) as FeatureCollection;
+      } else if (
+        (tmpData = jsonParseNoThrow(data) ?? gpxOrKml(data)) !== null
```

Comment:
I know you were going for less lines of code here, but this is definitely a footgun and hard to decipher. There's also a lot of clever avoidance of typescript going on overall in this function which is _fine_ if you're not creating lines like this.

Reply:
Updating for better readability then